### ISSUE TITLE
Key-value Data Marts

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#33581]
**Use case**

ClickHouse should be able to serve large amount of short queries on prepared datasets.

For example, you can prepare a dataset for user personalization or antifraud and serve it directly by clickhouse-server in realtime in ad-network, RTB, for DDoS prevention, e-commerce apps, games, etc...

**Describe the solution you'd like**

1. clickhouse-server should have TCPHandler that will implement Redis or Memcached protocol.
2. Add `get` method to `IStorage` interface and `supportsGetRequests`. The method takes array of Fields and returns array of Fields.
3. Implement this method for `StorageDictionary`, `EmbeddedRocksDB`, `Join`, `Set` and (when it will be ready), `AggregatingMemory`. Then implement it for `Distributed` tables.

**Comment by UnamedRus** [comment]
It will be good to have some table engine, which will store data in memory, but persistent storage will be backed by any *MergeTree table.

So when MergeTree table will receive new insert or part by replication protocol, in memory representation will auto update it's own state.

For today neither of provided option give that:

1. Join table engine have questionable storage on disk, which indefinitely create new .bin files, this will eventually lead to heavy disk usage and slow server start. It's also doesn't support replication.
2. AggregatingMemory doesn't have persistent storage and doesn't support replication.
3. Dictionaries doesn't support all data types, it's also will wait up to LIFETIME seconds before update, which isn't really real time update.
4. EmbeddedRocksDB doesn't support replication.

**Comment by alexey-milovidov** [comment]
Also consider MergeTree table with in-memory parts.
It has write-ahead log that is rotated and parts are merged in background.

Downsides:
- pre-production status;
- index is still sparse - not good for key-value requests or for point updates;
- merge of in-memory parts to in-memory part is not supported for now;
- no point updates (upsert, replace);

**Comment by UnamedRus** [comment]
> index is still sparse - not good for key-value requests or for point updates;

Yeah, but it's not in-memory, always finalized prepared data, so it's also have its own big flaws.

**Comment by dankondr** [comment]
I take it

**Comment by chrisgoddard** [comment]
Any updates on this feature?

**Comment by ZlataMoore** [comment]
I want to take it as graduation project

**Comment by chrisgoddard** [comment]
I have some use cases if you're looking for more examples/ideas

**Comment by ZlataMoore** [comment]
@alexey-milovidov Good morning! I’ve submitted my bachelor thesis work for this ticket. Could you please write a review?

I implemented a read-only Redis-compatible endpoint in ClickHouse for key-value lookups — a fast path to prepared tables via IKeyValueEntity::getByKeys. It supports PING, QUIT, SELECT, GET, MGET, Redis DB → ClickHouse database/table/default_column mapping, String and UInt64 keys, error handling, unsupported schemas, and basic limits for request size, key count, and MGET size.

Results:
On a 100k String-key benchmark, p99 latency for a single lookup decreased from ~1.021 ms with HTTP SQL to ~0.021 ms with the Redis endpoint. Throughput increased from ~2.8k request QPS with HTTP SQL to ~84k command QPS with the Redis endpoint. Raw Redis reference performance was ~90k QPS.

I also tested on bench.kv_test with 10M rows and UInt64 keys: p99 latency for a single lookup decreased from ~1.059 ms to ~0.028 ms, and throughput increased from ~2.7k to ~80k QPS.

Additionally, I added integration tests, a demo UI showing the SQL path vs Redis endpoint path, as well as benchmark scripts and documentation.

Repository: (https://github.com/ZlataMoore/ClickHouse)

**Comment by vdimir** [comment]
@ZlataMoore that sounds awesome, thanks. Could you [create a PR](https://github.com/ClickHouse/ClickHouse/compare) in this repo, so we can start code review and discuss implementaion further?

**Comment by ZlataMoore** [comment]
Yes, of course, I will create a PR. We have a deadline today(May 20th) for submitting the supervisor reviews for our theses. Could you please let me know if you might be able to contact Alexey Milovidov and ask him to write the reviews for us?

**Comment by alexey-milovidov** [comment]
Contacted.

**Comment by kaijaytu** [comment]
Hi, this is listed in the Intern Tasks 2025/2026 (#87836). Is it still open for contribution? I'd like to take it on.
